oval/redhat/redhat_test.go

```diff
@@ -0,0 +1,146 @@
+			bzip2FileNames: map[string]string{
+				"/RHEL6/rhel-6.oval.xml.bz2": "testdata/test.txt",
+			},
+			goldenFiles:      map[string]string{},
+			expectedErrorMsg: "failed to unmarshal Red Hat OVAL XML: bzip2 data invalid: bad magic value",
+		},
+		{
+			name: "broken XML",
+			bzip2FileNames: map[string]string{
+				"/RHEL6/rhel-6.oval.xml.bz2": "testdata/rhel-6-broken.oval.xml.bz2",
+			},
+			goldenFiles:      map[string]string{},
+			expectedErrorMsg: "failed to unmarshal Red Hat OVAL XML: XML syntax error on line 42: unexpected EOF",
+		},
+		{
+			name: "invalid RHSA-ID is ignored",
+			bzip2FileNames: map[string]string{
+				"/RHEL6/rhel-6.oval.xml.bz2": "testdata/rhel-6-invalid-id.oval.xml.bz2",
+				"/RHEL7/rhel-7.oval.xml.bz2": "testdata/rhel-7.oval.xml.bz2",
+				"/RHEL8/rhel-8.oval.xml.bz2": "testdata/rhel-8.oval.xml.bz2",
+			},
+			goldenFiles: map[string]string{
+				"/tmp/oval/redhat/7/2015/RHBA-2015:0364.json": "testdata/golden/RHBA-2015:0364.json",
+				"/tmp/oval/redhat/7/2015/RHBA-2015:0386.json": "testdata/golden/RHBA-2015:0386.json",
+				"/tmp/oval/redhat/7/2015/RHBA-2015:0441.json": "testdata/golden/RHBA-2015:0441.json",
+
+				"/tmp/oval/redhat/8/2019/RHSA-2019:0966.json": "testdata/golden/RHSA-2019:0966.json",
+				"/tmp/oval/redhat/8/2019/RHSA-2019:0968.json": "testdata/golden/RHSA-2019:0968.json",
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				filePath, ok := tc.bzip2FileNames[r.URL.Path]
+				if !ok {
+					http.NotFound(w, r)
+					return
+				}
+				b, err := ioutil.ReadFile(filePath)
+				assert.NoError(t, err, tc.name)
+				_, err = w.Write(b)
+				assert.NoError(t, err, tc.name)
+			}))
+			defer ts.Close()
+
+			u := ts.URL + "/RHEL%s/rhel-%s.oval.xml.bz2"
+			fmt.Println(u)
+			appFs := afero.NewMemMapFs()
+			c := Config{
+				VulnListDir: "/tmp",
+				URLFormat:   u,
+				AppFs:       appFs,
+				Retry:       0,
+			}
+			err := c.Update()
+			switch {
+			case tc.expectedErrorMsg != "":
+				assert.Contains(t, err.Error(), tc.expectedErrorMsg, tc.name)
+			default:
+				assert.NoError(t, err, tc.name)
+			}
+
+			fileCount := 0
+			err = afero.Walk(appFs, "/", func(path string, info os.FileInfo, err error) error {
+				if err != nil {
+					return err
+				}
+				if info.IsDir() {
+					return nil
+				}
+				fileCount += 1
+
+				actual, err := afero.ReadFile(appFs, path)
+				assert.NoError(t, err, tc.name)
+
+				goldenPath, ok := tc.goldenFiles[path]
+				assert.True(t, ok, tc.name)
+				if *update {
+					err = ioutil.WriteFile(goldenPath, actual, 0666)
+					assert.NoError(t, err, tc.name)
+				}
+				expected, err := ioutil.ReadFile(goldenPath)
+				assert.NoError(t, err, tc.name)
+
+				assert.Equal(t, expected, actual, tc.name)
+
+				return nil
+			})
+			assert.Equal(t, len(tc.goldenFiles), fileCount, tc.name)
+			assert.NoError(t, err, tc.name)
+		})
```

Comment:
Can we factor this out to a separate integration test file? Keeping a unit test (the first half of this test) and the integration test together is confusing to read and hard to maintain.

Reply:
afero is used for FileSystem mocking. If you use NewMemMapFs, files are saved on a memory, not on a local file system. So, this is unit testing.
https://github.com/spf13/afero#step-2-declare-a-backend